### ISSUE TITLE
YouTube Search Feature Request: use Invidious/Newpipe API as an option?

### ISSUE THREAD

**Issue by MintMain21** [CogentRedTester/mpv-scripts#13]
I have succesfully made MPV into my personal dream player, thanks in no small part to plug-in developers like you. YouTube Search was a plugin I initially avoided because I didn't want to log into my Gmail Account just to fetch the required YT API, but one day I decided "to heck with it" because I do most of my YT Viewing through MPV anyway. So I opened a temporary container in FireFox, got my API Key, plugged it into this plugin, and I've been happily using it since.

....But here's the thing. What if I wanted to reccomend this plugin to someone who doesn't have a Google/GMail account, and is unwilling to make one just to use this plugin? What workaround could they use?

My proposal? Invidious API. Or Newpipe API. Either option works. I don't know if it could work with this plugin, but I can only presume it's possible because of the many apllications that offer to use Invidious/Newpipe API as well as Google API, including FreeTube, NewPipe (duh), etc.

**Comment by CogentRedTester** [comment]
Do you have a link to the Invidious/Newpipe API documentation?

**Comment by CogentRedTester** [comment]
Ok, I believe I have added support for invidious API calls. You need to set the new `invidious` option to `yes`. You also have to use the new `api_path` option to specify a responsive invidious public API. The paths seem to be in the form `https://invidiouspath.com/api/v1/`. You will probably also want to change the new `frontend` option, which is the url of the actual site you want to be trying to load the videos from. Set it to the url of whatever invidious frontend you're using.

**Comment by CogentRedTester** [comment]
This feature is currently in an experimental state, so please let me know about any bugs or strange behaviour. Once you have confirmed it is working we can close the issue.

**Comment by MintMain21** [comment]
Well, after toggling the appropriate inputs in the script, I have it set so that it will use an Invidious Instance to search for videos and load the YouTube URLS from those search results. So far so good. Pretty Much exactly how I wanted it.

One additional feature request that I'm going to suggest for quality-of-life purposes is the potential to fallback on another API if the chosen default were to fail (for example, using YouTube API if the Invidious option fails, or using a different Invidious Instance altogether). 

Other than that, everything is working. I'll leave you to decide when this issue should be closed. If any bugs pop up I'll keep you updated.

**Comment by CogentRedTester** [comment]
Try the latest commit. It should work but is completely untested.

**Comment by MintMain21** [comment]
Alright, so to test out the fallback API feature, I left the default as YouTube without an API Key, and set the fallback option to an invidious instance (Use Invidious API Calls is False, Use Invidious for Fallback is true). This didn't work. I tried again by setting both Invidious API and Invidious for Fallback as true, and this did not work either. I tried again by using Invidious for both the default and the fallback (with the defunct invidio.us domain deliberately set as the default), and this time it did work. Finally, I left the default as the defunct invidio.us, but set the fallback to be YouTube using a proper API Key. This does not work.

TL;DR: If both the default and the fallback are Invidious domains, this works fine. mix/mingling them with YouTube API however, does not. 

I'm not sure it's plausible to use an Invidious API as a fallback if the default is YouTube nor visa-versa, but if all we want is Invidious Domain fallbacks for Invidious Domain Defaults, then yes, this works.

**Comment by CogentRedTester** [comment]
Can you try again now? I believe the last commit should have fixed those issues.

**Comment by MintMain21** [comment]
Everything's working now, including the mixing/matching of YouTube and Invidious API. 

It might be nice to have multiple layers of fallbacks (so, like, an additional fallback if the default and the first fallback fail), but otherwise this is everything I wanted. 

I'll leave you to close it if you feel like you've done enough.

**Comment by RafeeDaBoy** [comment]
@MintMain21 I was having some problems with setting up YouTube search. I placed the necessary scripts in their respective folders and the API key as well inside the youtube-search.lua file. But press Shift+Y or Ctrl+Y doesn't do anything. Am I missing anything? Thanks!

P.S I've been trying to perfect mpv and I'd love to know what plugins you're using!

**Comment by MintMain21** [comment]
Okay, @RafeeDaBoy , first lets get over some basic usage information.

First, you need the youtube-search.lua to be in the scripts folder, alongside scripts like autoload, and ontop-playback. So if those scripts work, then you have the right location.

Second of all, the user-imput.lua and scroll-list.lua both need to be in the script-modules folder. If this doesn't already exist, make it in the same directory where your scripts folder is. It should look something like this: 
![image](https://user-images.githubusercontent.com/91362684/151638179-fe9cffc1-6a71-47c3-835e-b4cb96041a5d.png)


Next, if you want the default behavior to be the YouTube API and Invidious as the backup, then (and make sure you're using a text-editing program for this, as in a coding program. Something like LibreOffice or Word won't be good enough) go to the same area where you typed out your API Key, make sure the default API path is https://www.googleapis.com/youtube/v3/ and then set the fallback Invidious API to https://[insert invidious domain here]//api/v1/. Finally, make sure "use invidious API calls" is set to False, but "fallback Invidious" is set to True. 

Any stable Invidious Instance that is publicly available and uses HTTPS for secure connections should work. From my experience, vid.puffyan.us is the most reliable instance. 

Bonus: If you leave "URL to load videos from" as the default value, then while Invidious is used to search for YouTube videos, YouTube URLS are what are loaded upon selection. (This of this like selecting the YouTube icond on the search page below a video). If you want, you can set this to whatever Invidious Instance you prefer, but I find YouTube URLS are easier to share.

As for other MPV Scripts? Well, some of my favorites include Bookmarker-Menu, Easycrop, File-Browser, Pause-Indicator, Playlistmanager, reload, SmartCopyPaste, SponsorBlock_Minimal, and StopCache.

**Comment by RafeeDaBoy** [comment]
Thanks a lot for the response. For some reason using the directory where my mpv config and scripts were didn't and putting it in ~AppData\Roaming\mpv data resolved it. 

As for the [Invidious instance](https://docs.invidious.io/Invidious-Instances.md), wouldn't choosing the one with region closest to me be the best option?

As for the bonus part, just to be sure, if I choose an Invidious instance for "URL to load videos from", it's gonna show the video thumbnails?

One more thing, when I select a channel from the list, instead of showing all the videos of that channel, it just shows a blank screen and plays that channel's latest video after a few minutes. Is that normal? And is there away to have the text of youtube-search show up in front of the mpv logo?
![image](https://user-images.githubusercontent.com/69535896/151659338-e4333a6d-9dba-4c6a-8caf-ae6a31724246.png)
![image](https://user-images.githubusercontent.com/69535896/151659206-8ba4afc1-9259-4c0e-b394-3a16e6190f1b.png)

**Comment by MintMain21** [comment]
You are correct that you should probably use an invidious instance closer to your geographic locaion. As a spoiled US American I personally find vid.puffyan.us to work the best. If another invidious instance works better for you, by all means use it.

Addmitidly, I've never had an issue with text being obscured by the MPV Logo, but both the lack of thumbnails and the default behavior when selecting a channel is normal. Actually, what happens when you select a channel is that it enqueues multiple videos based on recency, so if you use a plugin like Playlist Manager you could easily sift through a channel's recent uploads.

If @CogentRedTester would like to respond, he'd probably be more helpful.

As for me, I admititdly find the MPV plugin most helpful when I'm searching for specific videos. If I need to view video tumbnails or better navigate a channel's content, I either use a Internet Browser + Invidious, or Freetube. In either case, I can easily open videos from there into MPV with SmartCopyPaste.

**Comment by RafeeDaBoy** [comment]
I already had Playlist Manager and yeah, it shows up there thanks! 

Also, is it possible to have a random video play after the current one is done like it does on Invidious's website?

**Comment by MintMain21** [comment]
Autoplay? That would be a different plugin. https://github.com/cvzi/mpv-youtube-upnext

**Comment by CogentRedTester** [comment]
> Thanks a lot for the response. For some reason using the directory where my mpv config and scripts were didn't and putting it in ~AppData\Roaming\mpv data resolved it.
> 
> As for the [Invidious instance](https://docs.invidious.io/Invidious-Instances.md), wouldn't choosing the one with region closest to me be the best option?

You could do this, but I would imagine that the load on the server is more significant that geographic distance. Actually loading videos from that instance might be different, depending on how youtube-dl extracts the url.

> As for the bonus part, just to be sure, if I choose an Invidious instance for "URL to load videos from", it's gonna show the video thumbnails?

No, thumbnails are unsupported by this script, and never will be.

> One more thing, when I select a channel from the list, instead of showing all the videos of that channel, it just shows a blank screen and plays that channel's latest video after a few minutes. Is that normal?

When you load a channel this script just passes the channel url to mpv, which passes it to youtube-dl. I make no guarantees for how well that works, or what exactly gets added to the playlist, but currently it seems to add all the channel's videos sorted by latest uploaded. If there are a lot of those videos then youtube-dl may take a long time to parse the playlist.

> And is there away to have the text of youtube-search show up in front of the mpv logo?

That logo is drawn by the OSC when the player is idle. You can hide it by setting [osc-visibility](https://mpv.io/manual/master/#on-screen-controller-osc-visibility) to `never`. I believe by default the `del` key cycles this value.

**Comment by CogentRedTester** [comment]
Also these issues are really not the place for these discussions, please use the Discussions tab in the future.
